### ISSUE TITLE
LR scheduler lags behind after resume_from_checkpoint

### ISSUE THREAD

**Issue by voyleg** [PyTorchLightning/pytorch-lightning#4655]
## 🐛 Bug

MultiStepLR lags behind by one epoch after resuming from checkpoint in Trainer.
See this [collab](https://colab.research.google.com/drive/1Snq6QgPMfg3i5RzgtQxEFJTrmWlgYcFq?usp=sharing) to reproduce.

In the example below,
```python
class BoringModel(LightningModule):
    ...
    def training_step(self, batch, batch_idx):
        print(f'Epoch {self.trainer.current_epoch} / Step {self.trainer.global_step}: lr {self.trainer.optimizers[0].param_groups[0]["lr"]}')
        ...
    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=100)
        lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[4], gamma=.1)
        return [optimizer], [lr_scheduler]

def test_x(resume_from_checkpoint=None):
    model = BoringModel()
    trainer = pl.Trainer(
        callbacks=[pl.callbacks.ModelCheckpoint(save_top_k=-1)],
        resume_from_checkpoint=resume_from_checkpoint
    )
    trainer.fit(model, train)

test_x()
test_x('.../checkpoints/epoch=2.ckpt')
```
in the run from scratch the learning rate is changed at epoch 4 
```
Epoch 0 / Step 0: lr 100
Epoch 0 / Step 1: lr 100
Epoch 1 / Step 2: lr 100
Epoch 1 / Step 3: lr 100
Epoch 2 / Step 4: lr 100
Epoch 2 / Step 5: lr 100
Epoch 3 / Step 6: lr 100
Epoch 3 / Step 7: lr 100
Epoch 4 / Step 8: lr 10.0  # lr changed
Epoch 4 / Step 9: lr 10.0
Epoch 5 / Step 10: lr 10.0
Epoch 5 / Step 11: lr 10.0
```
and in the resumed run the learning rate is changed at epoch 5, i.e., the scheduler lags by one epoch.
```
Epoch 3 / Step 6: lr 100
Epoch 3 / Step 7: lr 100
Epoch 4 / Step 8: lr 100
Epoch 4 / Step 9: lr 100
Epoch 5 / Step 10: lr 10.0  # lr changed
Epoch 5 / Step 11: lr 10.0
```

### Environment
The output from the environment collection script:
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.0.6
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

**Comment by Borda** [comment]
@voyleg mind just for the record write here how much it is laging behind?

**Comment by voyleg** [comment]
It is lagging by one epoch. I added an example in the description.

**Comment by SkafteNicki** [comment]
If you print the `state_dict` of the scheduler in `training_step` you would see that:
first run
```
...
Epoch 2 / Step 5: lr 100
{'milestones': Counter({4: 1}), 'gamma': 0.1, 'base_lrs': [100], 'last_epoch': 2, '_step_count': 3, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [100]}
Epoch 3 / Step 6: lr 100
{'milestones': Counter({4: 1}), 'gamma': 0.1, 'base_lrs': [100], 'last_epoch': 3, '_step_count': 4, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [100]}
...
```
second run
```
Epoch 3 / Step 6: lr 100
{'milestones': Counter({4: 1}), 'gamma': 0.1, 'base_lrs': [100], 'last_epoch': 2, '_step_count': 3, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [100]}
Epoch 3 / Step 7: lr 100
{'milestones': Counter({4: 1}), 'gamma': 0.1, 'base_lrs': [100], 'last_epoch': 2, '_step_count': 3, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [100]}
...
```
as you can see the `state_dict` of the learning rate scheduler gets saved **before** it is stepped, meaning that it is the `state_dict` of the last epoch that is getting loaded on the second run.
It is these lines of codes that are the problem:
https://github.com/PyTorchLightning/pytorch-lightning/blob/e04e7c9eccc1233d6059e00e427b776cb9bf852a/pytorch_lightning/trainer/trainer.py#L520-L530
the call to `self.optimizer_connector.update_learning_rates(interval='epoch')` should happen before the checkpoint is saved, but since it happens somewhere inside the call to `self.train_loop.run_training_epoch()` this is not the case right now.

**Comment by edenlightning** [comment]
TODO: check the lr_schedular is using the right arguments in this test

**Comment by tchaton** [comment]
Hey @rohitgr7, any updates ?

**Comment by rohitgr7** [comment]
It's still WIP. Need some time.

**Comment by seannz** [comment]
This problem really got me today when I tried to resume training. Wouldn't the simplest fix be to call `update_learning_rates` before `run_training_epoch`? This is a perfectly valid training paradigm. See below:

> Prior to PyTorch 1.1.0, the learning rate scheduler was expected to be called before the optimizer’s update; 1.1.0 changed this behavior in a BC-breaking way. If you use the learning rate scheduler (calling scheduler.step()) before the optimizer’s update (calling optimizer.step()), this will skip the first value of the learning rate schedule. If you are unable to reproduce results after upgrading to PyTorch 1.1.0, please check if you are calling scheduler.step() at the wrong time.

**Comment by rohitgr7** [comment]
> This problem really got me today when I tried to resume training. Wouldn't the simplest fix be to call `update_learning_rates` before `run_training_epoch`? This is a perfectly valid training paradigm. See below:
> 
> > Prior to PyTorch 1.1.0, the learning rate scheduler was expected to be called before the optimizer’s update; 1.1.0 changed this behavior in a BC-breaking way. If you use the learning rate scheduler (calling scheduler.step()) before the optimizer’s update (calling optimizer.step()), this will skip the first value of the learning rate schedule. If you are unable to reproduce results after upgrading to PyTorch 1.1.0, please check if you are calling scheduler.step() at the wrong time.

@seannz mind share the link for this statement??

**Comment by hyw1994** [comment]
@rohitgr7 You can find this when running a pure PyTorch project. Also, it is at: https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate.

**Comment by rohitgr7** [comment]
ok what you are saying will be like: (with init lr = 1e-2 with a decay of 0.1 per epoch)
epoch = 1
1. do first fit -> lr=1e-2
2. scheduler.step() -> lr=1e-2
3. optimizer.step() -> lr=1e-2
4. epoch done -> lr=1e-2
5. save checkpoint with scheduler -> lr=1e-2
6. resume checkpoint and load scheduler state -> lr=1e-2
7. run epoch -> lr=1e-2

after resuming training and doing .fit, it will run for lr=1e-2 in the second epoch. Did I miss anything here?

**Comment by hyw1994** [comment]
@rohitgr7 

I think that's the document means. Also, I noticed another problem when loading the saved checkpoints. To reproduce, you can start with the collab @seannz shared above. 

What you need to do is:

1. Change the milestone to 3
2. Run the test
3. Load the 4th checkpoint (any above the milestone)

You can see the learning rate is back to the initial value. I tried with 1.1.1 in colab and the behaviour is still the same.

**Comment by hyw1994** [comment]
@rohitgr7 @seannz 
I tried many ways, only one temporary fix could work:

1. Only load the model with model.load_state_dict() method. 
2. Update trainer global_step and epoch from saved checkpoint.
3. At configure_optimizers, call scheduler.step() epoch times.

The code is:

checkpoint = torch.load("./checkpoints/4.ckpt")
model.load_state_dict(checkpoint["state_dict"])
trainer = pl.Trainer(max_epochs=10, progress_bar_refresh_rate=20)
trainer.global_step = checkpoint["global_step"]
trainer.current_epoch = checkpoint["epoch"]

def configure_optimizers(self):
      # optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
      optimizer = torch.optim.SGD(self.parameters(), lr=100)
      lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[2, 6], gamma=.1)
      for _ in range(self.trainer.current_epoch):
          lr_scheduler.step()
      return [optimizer], [lr_scheduler]

**Comment by rohitgr7** [comment]
@hyw1994 working on a fix #5208, will try to fix this issue.

**Comment by hyw1994** [comment]
@rohitgr7 I saw the pull request, thanks for your help. This temporary code is just for people who need this urgently (like me).
